lib/src/components/NxFileUpload/NxMultiFileUpload/NxMultiFileUpload.tsx

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2019-present Sonatype, Inc.
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
+ */
+import React, { FormEvent, forwardRef, useRef, useContext, useEffect } from 'react';
+import { faExclamationCircle } from '@fortawesome/free-solid-svg-icons';
+import classnames from 'classnames';
+import { without } from 'ramda';
+
+import SelectedFile from '../SelectedFile';
+import NxFontAwesomeIcon from '../../NxFontAwesomeIcon/NxFontAwesomeIcon';
+import NxButton from '../../NxButton/NxButton';
+import { FormAriaContext } from '../../NxForm/context';
+import { useUniqueId } from '../../../util/idUtil';
+import { Props, propTypes, MultiSelectedFileProps } from '../types';
+
+export { Props };
+
+import './NxMultiFileUpload.scss';
+
+function SelectedFileWrapper({file, onDismiss: onDismissProp}: MultiSelectedFileProps) {
+  const fileRef = useRef<HTMLSpanElement>(null);
+
+  function onDismiss() {
+    onDismissProp(file);
+  }
+
+  useEffect(function() {
+    if (fileRef.current) {
+      fileRef.current.scrollIntoView({block: 'nearest'});
+    }
+  }, []);
+
+  return (
+    <SelectedFile file={file} onDismiss={onDismiss} ref={fileRef}/>
+  );
+}
+
+const NxMultiFileUpload = forwardRef<HTMLDivElement, Props>(function NxMultiFileUpload(props, ref) {
+  const {
+        className: classNameProp,
+        onChange: onChangeProp,
+        files,
+        isRequired,
+        isPristine,
+        id,
+        disabled,
+        ...attrs
+      } = props,
+      isFileSelected = !!files?.length,
+      { showValidationErrors: formShowValidationErrors } = useContext(FormAriaContext),
+      showValidationErrors = formShowValidationErrors || !isPristine,
+      showError = isRequired && showValidationErrors && !isFileSelected,
+      className = classnames('nx-multi-file-upload', 'nx-file-upload', classNameProp),
+      noFileMessageClassName = classnames('nx-file-upload__no-file-message', {
+        'nx-file-upload__no-file-message--invalid': showError
+      }),
+      inputRef = useRef<HTMLInputElement>(null),
+      inputId = useUniqueId('nx-multi-file-upload-input', id),
+      selectedFilesContainerRef = useRef<HTMLDivElement>(null),
+      validationErrorId = useUniqueId('nx-multi-file-upload-validation-error');
+
+  // DataTransfer has the only known API in which FileLists can be programatically constructed from various Files.
+  const dataTransferObject = new DataTransfer();
+
+  function manageFiles(selectedFiles: File[]) {
```

Comment:
This isn't really what I meant.  That this calls `onChangeProp`, and that this is itself called twice in `onChange`, means that `onChangeProp` is being called twice when `onChange` is called.  Also maintaining the `dataTransferObject`, a stateful object which is used and re-used in a mutable way, per render call is super dangerous and likely to have weird issues where it doesn't work.

What I meant for you to do is to create a function with a signature like 
```
function combineFileLists(...fileLists: ArrayLike<File>[]): FileList
```
which would instantiate a `DataTransfer` internally and use it to group together all of the files from the passed in `fileLists`, and then return that `DataTransfer`'s files.  This function would be functionally-pure - calling `onChangeProp` would occur outside of it.

Reply:
This function would be functionally-pure - calling `onChangeProp` would occur outside of it.

I updated the function here, and hopefully on the right track: [4a8d2b4](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/4a8d2b4fe09a940e49188fc6b1f156f1717f6453).  I still convert the FileList into an array inside `onDismiss` in order to remove `fileObj` from the array before using the `DataTransfer` object. If you'd rather me keep it as a  FIleList, I can look into other options as well.

Reply:
> I updated the function here, and hopefully on the right track: [4a8d2b4](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/4a8d2b4fe09a940e49188fc6b1f156f1717f6453). I still convert the FileList into an array inside onDismiss in order to remove fileObj from the array before using the DataTransfer object. If you'd rather me keep it as a FIleList, I will look into other options

Yes that's exactly what I meant. Don't forget to check the `length` of the resulting `FileList` in `onChange` so that we can keep our contract of returning `null` rather than an empty `FileList`.

Reply:
> then fixed missing ! here: [dfd2346](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/dfd2346acc8391d0b6e6236d6ffaf129c6f8df10)

One more improvement on that: when you have a conditional with both an if and an else (or both sides of a ternary) it's typically preferred to order the two sides such that the condition is positive.  i.e. in this case `returnedFileList.length ? returnedFileList : null`

Reply:
> One more improvement on that: when you have a conditional with both an if and an else (or both sides of a ternary) it's typically preferred to order the two sides such that the condition is positive. i.e. in this case `returnedFileList.length ? returnedFileList : null`

Fixed here: [7f73f92](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/7f73f92d99cce9cd2e40e17878a1a7b275287e1c)